CCIP: tutorial on estimating the gas limit (#1816)
smartcontractkit/documentation: src/content/ccip/tutorials/ccipreceive-gaslimit.mdx

1. Click on the debugger and you'll get the gas details:

   ```text
   "gas":{
   "gas_left":5685
   "gas_used":5031
   "total_gas_used":7994315
   }
   ```

1. Note the `gas_left` is equal to the limit that is set in the `sendCCIPMessages.ts` script: `5685`. The `gas_used` is the actual gas used by the Receiver contract to process the message.

1. Repeating the same steps for the other two messages, we can summarize the output:

   | Number of iterations | Estimated gas usage (Hardhat) | Buffered gas limit (+10%) | Gas used on testnet |
   | -------------------- | ----------------------------- | ------------------------- | ------------------- |
   | 0                    | 5168                          | 5685                      | 5031                |
   | 50                   | 14718                         | 16190                     | 14581               |
   | 99                   | 24077                         | 26485                     | 23940               |

> Testing on testnets has confirmed that a gas limit of 26,485 is adequate for the `ccipReceive` function to execute successfully under various conditions. It is important to note, however, that gas usage may differ across testnets. Therefore, conducting similar validation efforts on the blockchain where you intend to deploy is advisable. Deploying and validating contracts across multiple testnets can be time-consuming. For efficiency, consider utilizing offchain methods (see the next section) to estimate gas usage.

Reviewer: ```suggestion
Testing on testnets has confirmed that a gas limit of 26,485 is adequate for the `ccipReceive` function to execute successfully under various conditions. However, it is important to note that gas usage may differ across testnets. Therefore, it is advisable to conduct similar validation efforts on the blockchain where you intend to deploy. Deploying and validating contracts across multiple testnets can be time-consuming. For efficiency, consider utilizing offchain methods (see the next section) to estimate gas usage.
```

## Offchain methods

This section guides you through estimating gas usage using two different offchain methods:

> - A web3 provider using the [ethers.js `estimateGas`](https://docs.ethers.org/v6/api/providers/#Provider-estimateGas) function.

Reviewer: ```suggestion
- A Web3 provider using the [ethers.js `estimateGas`](https://docs.ethers.org/v6/api/providers/#Provider-estimateGas) function.
```

> - [Tenderly simulation API](https://docs.tenderly.co/reference/api#tag/Simulations/operation/simulateTransaction). The Tenderly simulation API provides a more accurate result (Read this [blog](https://blog.tenderly.co/how-tenderly-enables-most-accurate-ethereum-gas-estimation/) to learn more) but you are limited to the blockchains supported by Tenderly.

Reviewer: ```suggestion
- [Tenderly simulation API](https://docs.tenderly.co/reference/api#tag/Simulations/operation/simulateTransaction). The Tenderly simulation API provides a more accurate result (Read this [blog post](https://blog.tenderly.co/how-tenderly-enables-most-accurate-ethereum-gas-estimation/) to learn more) but you are limited to the blockchains supported by Tenderly.
```

These methods provide the most accurate and rapid means to determine the necessary gas limit for the `ccipReceive` function. You will use the same CCIP Receiver contract deployed on the Ethereum Sepolia testnet in the previous section.

### Prerequisites

> 1. In your terminal, change to the `offchain` directory:

Reviewer: ```suggestion
1. In your terminal, navigate to the `offchain` directory:
```

   ```bash

>    cd offchain

Reviewer: ```suggestion
   cd ../offchain
```

   ```

> 1. Modify the `data.json` file and put the deployed addresses of your Sender and Receiver contracts.

Reviewer: ```suggestion
1. Modify the `data.json` file to insert the deployed addresses of your Sender and Receiver contracts.
```

1. Install the dependencies:

   ```bash
   npm install
   ```

1. Set the password to encrypt your environment variables:

   ```bash
   npx env-enc set-pw
   ```

> 1. Setup these environment variables (**Remark**: Read the [tenderly guide](https://docs.tenderly.co/account/projects/account-project-slug) for finding the `TENDERLY_ACCOUNT_SLUG` and `TENDERLY_PROJECT_SLUG`. You can get `TENDERLY_ACCESS_KEY` by generating a new access token):

Reviewer: ```suggestion
1. Setup these environment variables (**Remark**: Read the [tenderly guide](https://docs.tenderly.co/account/projects/account-project-slug) for finding `TENDERLY_ACCOUNT_SLUG` and `TENDERLY_PROJECT_SLUG`. You can get `TENDERLY_ACCESS_KEY` by generating a new access token):
```

> ts-node src/estimateGasProvider.ts

Final Gas Usage Report:
Number of iterations 0 - Gas used: 5377
Number of iterations 50 - Gas used: 14946
Number of iterations 99 - Gas used: 24324
```

### Estimate gas using Tenderly

To estimate the gas usage of the `ccipReceive` function using Tenderly, execute the following command:

```bash
npm run estimate-gas-tenderly
```

Example output:

```text
$ npm run estimate-gas-tenderly

> offchain-simulator@1.0.0 estimate-gas-tenderly
> ts-node src/estimateGasTenderly.ts

Final Gas Usage Report:
Number of iterations 0 - Gas used: 5031
Number of iterations 50 - Gas used: 14581
Number of iterations 99 - Gas used: 23940
```

### Comparison

The table below summarizes the gas estimations for different iterations using both ethers.js and Tenderly:

| Number of iterations | Gas estimated (ethers.js) | Gas estimated (Tenderly) |
| -------------------- | ------------------------- | ------------------------ |
| 0                    | 5377                      | 5031                     |
| 50                   | 14946                     | 14581                    |
| 99                   | 24324                     | 23940                    |

> The gas estimations from both ethers.js and Tenderly are consistent across different iterations and are in line with actual testnet results. This demonstrates the accuracy and reliability of these offchain methods in estimating gas usage. However you can notice that Tenderly provides more accurate results.

Reviewer: ```suggestion
The gas estimations from both ethers.js and Tenderly are consistent across different iterations and align with actual testnet results. This demonstrates the accuracy and reliability of these offchain methods in estimating gas usage. However, you can notice that Tenderly provides more accurate results.
```

## Conclusion

> This tutorial has guided you through estimating the gas limit for the `ccipReceive` function using various methods. You have learned how to estimate gas usage in a local environment using Hardhat and Foundry, validate these estimations on testnets, and utilize offchain methods to estimate gas usage. The offchain methods, particularly Tenderly, have proven to be the most accurate and reliable in estimating gas usage. By following these methods, you can accurately determine the necessary gas limit for the `ccipReceive` function, ensuring cost-effective CCIP transactions. It is important to note that gas usage may differ across different blockchains. Therefore, conducting similar validation efforts on the blockchain where you intend to deploy is advisable. For efficiency, consider utilizing offchain methods to estimate gas usage.

Reviewer: ```suggestion
This tutorial has guided you through estimating the gas limit for the `ccipReceive` function using various methods. You have learned how to estimate gas usage in a local environment using Hardhat and Foundry, validate these estimations on testnets, and utilize offchain methods to estimate gas usage.

The offchain methods, particularly Tenderly, have proven to be the most accurate and reliable in estimating gas usage. Following these methods, you can accurately determine the necessary gas limit for the `ccipReceive` function, ensuring cost-effective CCIP transactions. It is important to note that gas usage may differ across different blockchains. Therefore, conducting similar validation efforts on the blockchain where you intend to deploy is advisable. For efficiency, consider utilizing offchain methods to estimate gas usage.
```

1. **Iteration:** The function executes a loop from 0 to `maxIterations`, simulating variable computational work based on the input data. This variability directly influences gas consumption.
1. **Event Emission:** Finally, an event `MessageReceived` is emitted.

This code shows how gas consumption for the `_ccipReceive` function can fluctuate in response to the input data, highlighting the necessity for thorough testing under different scenarios to determine the correct `gasLimit`.

## Gas estimation in a local environment

To facilitate testing within a local environment, you will use the [MockCCIPRouter](https://github.com/smartcontractkit/ccip/blob/ccip-develop/contracts/src/v0.8/ccip/test/mocks/MockRouter.sol) contract. This contract serves as a mock implementation of the CCIP Router contract, enabling the local testing of CCIP Sender and Receiver contracts.
A notable feature of the `MockCCIPRouter` contract is its ability to emit a `MsgExecuted` event:

```solidity
event MsgExecuted(bool success, bytes retData, uint256 gassed))
```

This event reports the amount of gas consumed by the `ccipReceive` function.

### Foundry

#### Prerequisites

> 1. In tour terminal, change to the `foundry` directory:

Reviewer 2: ```suggestion
1. In your terminal, change to the `foundry` directory:
```

   ```

1. Open the [CCIP explorer](https://ccip.chain.link), search each message by its ID, and wait for each message to be successfully transmitted (`Status` in the explorer: `Success`).

For the example above, here are the destination transaction hashes:

    | Message id                                                         | Ethereum Sepolia transaction hash                                  |
    | ------------------------------------------------------------------ | ------------------------------------------------------------------ |
    | 0xf23b17366d69159ea7d502835c4178a1c1d1d6325edf3d91dca08f2c7a2900f7 | 0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689 |
    | 0x4b3a97f6ac959f67d769492ab3e0414e87fdd9c143228f9c538b22bb695ca728 | 0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689 |
    | 0x37d1867518c0f8c54ceb0c5507b46b8d44c6c53864218f448cba0234f8de867a | 0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689 |

**Note** that the Ethereum Sepolia transaction hash is the same for all the messages. This is because CCIP batched the messages.

### Check the actual gas usage

> 1. Open [Tenderly](https://dashboard.tenderly.co) and Search for the [destination transaction hash](https://dashboard.tenderly.co/tx/sepolia/0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689).

Reviewer 2: ```suggestion
1. Open [Tenderly](https://dashboard.tenderly.co) and search for the [destination transaction hash](https://dashboard.tenderly.co/tx/sepolia/0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689).
```

> 1. Search for `_callWithExactGasSafeReturnData` with a payload containing your messageId (without `0x`). [Example](https://dashboard.tenderly.co/tx/sepolia/0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689?trace=0.8.4.2) for `0xf23b17366d69159ea7d502835c4178a1c1d1d6325edf3d91dca08f2c7a2900f7`.

Reviewer 2: ```suggestion
1. Search for `_callWithExactGasSafeReturnData` with a payload containing your `messageId` (without `0x`). [Example](https://dashboard.tenderly.co/tx/sepolia/0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689?trace=0.8.4.2) for `0xf23b17366d69159ea7d502835c4178a1c1d1d6325edf3d91dca08f2c7a2900f7`.
```

> 1. Below it, you will find the call trace from the Router to your Receiver contract.[Call trace example](https://dashboard.tenderly.co/tx/sepolia/0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689?trace=0.8.4.2.0).

Reviewer 2: ```suggestion
1. Below the payload with your `messageId`, you will find the call trace from the Router to your Receiver contract. [Call trace example](https://dashboard.tenderly.co/tx/sepolia/0xf004eb6dab30b3cfb9d1d631c3f9832410b8d4b3179e65b85730563b67b1e689?trace=0.8.4.2.0).
```

> 1. Click on the debugger and you'll get the gas details:

Reviewer 2: Assuming you mean this Debugger tab?

<img width="227" alt="Screenshot 2024-03-18 at 1 20 14 AM" src="https://github.com/smartcontractkit/documentation/assets/5498502/758f43ff-e149-4ab3-b382-cab2d3f83bb7">

Not suggesting we add screenshots of Tenderly UI here, but we should refer to UI elements very precisely for clarity and accessibility.

- A Web3 provider using the [ethers.js `estimateGas`](https://docs.ethers.org/v6/api/providers/#Provider-estimateGas) function.
- [Tenderly simulation API](https://docs.tenderly.co/reference/api#tag/Simulations/operation/simulateTransaction). The Tenderly simulation API provides a more accurate result (Read this [blog post](https://blog.tenderly.co/how-tenderly-enables-most-accurate-ethereum-gas-estimation/) to learn more) but you are limited to the blockchains supported by Tenderly.

These methods provide the most accurate and rapid means to determine the necessary gas limit for the `ccipReceive` function. You will use the same CCIP Receiver contract deployed on the Ethereum Sepolia testnet in the previous section.

### Prerequisites

1. In your terminal, navigate to the `offchain` directory:

   ```bash
   cd ../offchain
   ```

1. Modify the `data.json` file to insert the deployed addresses of your Sender and Receiver contracts.

1. Install the dependencies:

   ```bash
   npm install
   ```

1. Set the password to encrypt your environment variables:

   ```bash
   npx env-enc set-pw
   ```

> 1. Setup these environment variables (**Remark**: Read the [tenderly guide](https://docs.tenderly.co/account/projects/account-project-slug) for finding `TENDERLY_ACCOUNT_SLUG` and `TENDERLY_PROJECT_SLUG`. You can get `TENDERLY_ACCESS_KEY` by generating a new access token):

Reviewer 2: ```suggestion
1. Setup these environment variables (**Remark**: Read the [Tenderly guide](https://docs.tenderly.co/account/projects/account-project-slug) for finding `TENDERLY_ACCOUNT_SLUG` and `TENDERLY_PROJECT_SLUG`. You can get `TENDERLY_ACCESS_KEY` by generating a new access token):
```

    - `ETHEREUM_SEPOLIA_RPC_URL`: The RPC URL for Ethereum Sepolia testnet. You can sign up for a personal endpoint from [Alchemy](https://www.alchemy.com/), [Infura](https://www.infura.io/), or another node provider service.
    - `AVALANCHE_FUJI_RPC_URL`: The RPC URL for Avalanche Fuji testnet. You can sign up for a personal endpoint from [Infura](https://www.infura.io/) or another node provider service.
    - `ETHERSCAN_API_KEY`: An Ethereum explorer API key, used to verify your contract. Follow [this guide](https://docs.etherscan.io/getting-started/viewing-api-usage-statistics) to get one from Etherscan.

    Input these variables using the following command:

           ```bash
           npx env-enc set
           ```

1.  Compile the contracts:

    ```bash
    npx hardhat compile
    ```

    The output should be similar to:

    ```text
    Generating typings for: 31 artifacts in dir: typechain-types for target: ethers-v6
    Successfully generated 114 typings!
    Compiled 33 Solidity files successfully (evm target: paris).
    ```

#### Estimate gas

> Located in the `test` directory, the `Send-Receive.ts` test file is designed to evaluate the gas usage of the `ccipReceive` function. This file employs the same logic as the Foundry test file, featuring three scenarios varying by the number of iterations. The test case transmits a CCIP message to the `MockCCIPRouter` contract, triggering the `MsgExecuted` event. This event provides insights into the gas requirements of the `ccipReceive` function by detailing the amount of gas utilized.

Reviewer 3: ```suggestion
Located in the `test` directory, the `Send-Receive.ts` test file is designed to evaluate the gas usage of the `ccipReceive` function. This file employs the same logic as the Foundry test file, featuring three scenarios varying by the number of iterations. The test case transmits a CCIP message to the `MockCCIPRouter` contract, triggering the `MsgExecuted` event. This event provides insights into the gas requirements of the `ccipReceive` function by detailing the amount of gas used.
```

To deploy and configure the CCIP Sender contract on the Avalanche Fuji testnet and the CCIP Receiver contract on the Ethereum Sepolia testnet, follow the steps below. **Note**: Your account must have some ETH tokens on Ethereum Sepolia and AVAX tokens on Avalanche Fuji.

1. Deploy the CCIP Sender on the Avalanche Fuji testnet:

   ```bash
   npx hardhat run scripts/deployment/deploySender.ts --network avalancheFuji
   ```

1. Deploy the CCIP Receiver on the Ethereum Sepolia testnet:

   ```bash
   npx hardhat run scripts/deployment/deployReceiver.ts --network ethereumSepolia
   ```

1. Authorize the Sender to send messages to Ethereum Sepolia:

   ```bash
   npx hardhat run scripts/configuration/allowlistingForSender.ts --network avalancheFuji
   ```

1. Authorize the Receiver to receive messages from the Sender:

   ```bash
   npx hardhat run scripts/configuration/allowlistingForReceiver.ts --network ethereumSepolia
   ```

Upon completion, you will find the CCIP Sender and Receiver contracts deployed and configured on their respective testnets. Contract addresses are available in the `scripts/generatedData.json` file.

### Send CCIP Messages

> 1. Send three CCIP messages with different number of iterations:

Reviewer 2: ```suggestion
1. Send three CCIP messages with different numbers of iterations:
```

1. Click on the _Debugger_ tab and you'll get the gas details:

   ```text
   "gas":{
   "gas_left":5685
   "gas_used":5031
   "total_gas_used":7994315
   }
   ```

1. Note the `gas_left` is equal to the limit that is set in the `sendCCIPMessages.ts` script: `5685`. The `gas_used` is the actual gas used by the Receiver contract to process the message.

1. Repeating the same steps for the other two messages, we can summarize the output:

   | Scenario                 | Number of iterations | Estimated gas usage (Hardhat) | Buffered gas limit (+10%) | Gas used on testnet |
   | ------------------------ | -------------------- | ----------------------------- | ------------------------- | ------------------- |
   | Baseline gas consumption | 0                    | 5168                          | 5685                      | 5031                |
   | Average gas consumption  | 50                   | 14718                         | 16190                     | 14581               |
   | Peak gas consumption     | 99                   | 24077                         | 26485                     | 23940               |

> Testing on testnets has confirmed that a gas limit of 26,485 is adequate for the `ccipReceive` function to execute successfully under various conditions. However, it is important to note that gas usage may differ across testnets. Therefore, it is advisable to conduct similar validation efforts on the blockchain where you intend to deploy. Deploying and validating contracts across multiple testnets can be time-consuming. For efficiency, consider utilizing offchain methods (see the next section) to estimate gas usage.

Reviewer 2: ```suggestion
Testing on testnets has confirmed that a gas limit of 26,485 is adequate for the `ccipReceive` function to execute successfully under various conditions. However, it is important to note that gas usage may differ across testnets. Therefore, it is advisable to conduct similar validation efforts on the blockchain where you intend to deploy. Deploying and validating contracts across multiple testnets can be time-consuming. For efficiency, consider using [offchain methods](#offchain-methods) to estimate gas usage.
```

- A Web3 provider using the [ethers.js `estimateGas`](https://docs.ethers.org/v6/api/providers/#Provider-estimateGas) function.
- [Tenderly simulation API](https://docs.tenderly.co/reference/api#tag/Simulations/operation/simulateTransaction). The Tenderly simulation API provides a more accurate result (Read this [blog post](https://blog.tenderly.co/how-tenderly-enables-most-accurate-ethereum-gas-estimation/) to learn more) but you are limited to the blockchains supported by Tenderly.

These methods provide the most accurate and rapid means to determine the necessary gas limit for the `ccipReceive` function. You will use the same CCIP Receiver contract deployed on the Ethereum Sepolia testnet in the previous section.

### Prerequisites

1. In your terminal, navigate to the `offchain` directory:

   ```bash
   cd ../offchain
   ```

1. Modify the `data.json` file to insert the deployed addresses of your Sender and Receiver contracts.

1. Install the dependencies:

   ```bash
   npm install
   ```

1. Set the password to encrypt your environment variables:

   ```bash
   npx env-enc set-pw
   ```

> 1. Setup these environment variables (**Remark**: Read the [Tenderly guide](https://docs.tenderly.co/account/projects/account-project-slug) for finding `TENDERLY_ACCOUNT_SLUG` and `TENDERLY_PROJECT_SLUG`. You can get `TENDERLY_ACCESS_KEY` by generating a new access token):
> 
>    - `ETHEREUM_SEPOLIA_RPC_URL`: The RPC URL for Ethereum Sepolia testnet. You can sign up for a personal endpoint from [Alchemy](https://www.alchemy.com/), [Infura](https://www.infura.io/), or another node provider service.
>    - `TENDERLY_ACCOUNT_SLUG`.
>    - `TENDERLY_PROJECT_SLUG`.
>    - `TENDERLY_ACCESS_KEY`.

Reviewer 2: ```suggestion
1. Set up the following environment variables:

   - `ETHEREUM_SEPOLIA_RPC_URL`: The RPC URL for Ethereum Sepolia testnet. You can sign up for a personal endpoint from [Alchemy](https://www.alchemy.com/), [Infura](https://www.infura.io/), or another node provider service.
   - `TENDERLY_ACCOUNT_SLUG`: This is one part of your Tenderly API URL. You can [find this value in your Tenderly account](https://docs.tenderly.co/account/projects/account-project-slug).
   - `TENDERLY_PROJECT_SLUG`: This is one part of your Tenderly API URL. You can [find this value in your Tenderly account](https://docs.tenderly.co/account/projects/account-project-slug).
   - `TENDERLY_ACCESS_KEY`: If you don't already have one, you can [generate a new access token](https://docs.tenderly.co/account/projects/how-to-generate-api-access-token).
```

   Input these variables using the following command:

   ```bash
   npx env-enc set
   ```

1. Generate [Typechain typings](https://www.npmjs.com/package/typechain) for the Receiver contract:

   ```bash
   npm run generate-types
   ```

### Introduction of the scripts

The scripts are located in the `src` directory. Each script is self-explanatory and includes comprehensive comments to explain its functionality and usage. There are three scripts:

> - `estimateGasProvider.ts`: This script utilizes the [`eth_estimateGas`](https://docs.alchemy.com/reference/eth-estimategas) Ethereum API to estimate the gas usage of the `ccipReceive` function. It simulates sending three CCIP messages to the Receiver contract with a varying number of iterations and estimates the gas usage using the [ethers.js `estimateGas`](https://docs.ethers.org/v6/api/providers/#Provider-estimateGas) function.

Reviewer 3: ```suggestion
- `estimateGasProvider.ts`: This script uses the [`eth_estimateGas`](https://docs.alchemy.com/reference/eth-estimategas) Ethereum API to estimate the gas usage of the `ccipReceive` function. It simulates sending three CCIP messages to the Receiver contract with a varying number of iterations and estimates the gas usage using the [ethers.js `estimateGas`](https://docs.ethers.org/v6/api/providers/#Provider-estimateGas) function.
```

> ts-node src/estimateGasTenderly.ts

Final Gas Usage Report:
Number of iterations 0 - Gas used: 5031
Number of iterations 50 - Gas used: 14581
Number of iterations 99 - Gas used: 23940
```

### Comparison

The table below summarizes the gas estimations for different iterations using both Web3 provider and Tenderly:

| Scenario                 | Number of iterations | Gas estimated (Web3 provider) | Gas estimated (Tenderly) |
| ------------------------ | -------------------- | ----------------------------- | ------------------------ |
| Baseline gas consumption | 0                    | 5377                          | 5031                     |
| Average gas consumption  | 50                   | 14946                         | 14581                    |
| Peak gas consumption     | 99                   | 24324                         | 23940                    |

The gas estimations from both Web3 provider and Tenderly are consistent across different iterations and align with actual testnet results. This demonstrates the accuracy and reliability of these offchain methods in estimating gas usage. However, you can notice that Tenderly provides more accurate results.

## Conclusion

> This tutorial has guided you through estimating the gas limit for the `ccipReceive` function using various methods. You have learned how to estimate gas usage in a local environment using Hardhat and Foundry, validate these estimations on testnets, and utilize offchain methods to estimate gas usage.

Reviewer 3: ```suggestion
This tutorial has guided you through estimating the gas limit for the `ccipReceive` function using various methods. You have learned how to estimate gas usage in a local environment using Hardhat and Foundry, validate these estimations on testnets, and use offchain methods to estimate gas usage.
```

As we have explored various methods for estimating gas for the `ccipReceive` function, it is crucial to apply this knowledge effectively. Here are some targeted recommendations to enhance your approach to gas estimation:

1. **Comprehensive Testing:** Emphasize testing under diverse scenarios to ensure your gas estimations are robust. Different conditions can significantly affect gas usage, so covering as many cases as possible in your tests is crucial.
1. **Preliminary Local Estimates:** Local testing is a critical first step for estimating gas and ensuring your contracts function correctly under various scenarios. While Hardhat and Foundry facilitate development and testing, it's key to remember that these environments may not perfectly mirror your target blockchain's conditions. These initial estimates serve as valuable insights, guiding you toward more accurate gas limit settings when you proceed to testnet validations. Incorporating a buffer based on these preliminary results can help mitigate the risks of underestimating gas requirements due to environmental differences.

> 1. **Efficiency with Offchain Methods:** Recognizing that testing across different blockchains can be resource-intensive, leaning on offchain methods for gas estimation becomes invaluable. Tools such as Tenderly not only facilitate rapid and accurate gas usage insights on your target blockchains but also enable you to simulate the execution of the `ccipReceive` function on actual contracts deployed on mainnets. If Tenderly doesn't support a particular blockchain, leveraging a Web3 provider that does, as illustrated in the [Estimate gas using a Web3 provider](#estimate-gas-using-a-web3-provider) section, becomes a practical alternative. This is particularly helpful when considering the diversity in gas metering rules across blockchains. This approach saves time and enhances the precision of your gas limit estimations, allowing for more cost-effective transactions from your dApp.

Reviewer 2: ```suggestion
1. **Efficiency with Offchain Methods:** Since testing across different blockchains can be resource-intensive, leaning on offchain methods for gas estimation is invaluable. Tools such as Tenderly not only facilitate rapid and accurate gas usage insights on your target blockchains but also enable you to simulate the execution of the `ccipReceive` function on actual contracts deployed on mainnets. If Tenderly doesn't support a particular blockchain, a practical alternative is to use a Web3 provider that does support that chain, as illustrated in the [Estimate gas using a Web3 provider](#estimate-gas-using-a-web3-provider) section. This is particularly helpful when considering the diversity in gas metering rules across blockchains. This approach saves time and enhances the precision of your gas limit estimations, allowing for more cost-effective transactions from your dApp.
```

### Estimate gas usage on your local environment

Now that you've locally estimated the gas usage of the `ccipReceive` function using the provided projects, you can apply the same approach to your own Foundry or Hardhat project. This section will guide you through estimating gas usage in your Foundry or Hardhat project.

#### Estimate `ccipReceive` gas usage locally in your Foundry project

To estimate the gas usage of the `ccipReceive` function within your own Foundry project, follow these steps:

1. Create a testing file in the `test` directory of your project and import the `MockCCIPRouter` contract:

   ```solidity

>    import "@chainlink/contracts/src/v0.8/ccip/test/mocks/MockRouter.sol";

Reviewer: ```suggestion
   import {MockCCIPRouter} from "@chainlink/contracts-ccip/src/v0.8/ccip/test/mocks/MockRouter.sol";
```

   1. After sending the CCIP message, use `vm.getRecordedLogs()` to collect the recorded logs. For more details, check this [example](https://github.com/smartcontractkit/smart-contract-examples/blob/4b3800967e475ecd990fd417bec9960dd9ebe2af/ccip/estimate-gas/foundry/test/SendReceive.t.sol#L47C32-L47C52).

   1. Parse the logs to find the `MsgExecuted(bool,bytes,uint256)` event and extract the gas used. For more details, check this [example](https://github.com/smartcontractkit/smart-contract-examples/blob/main/ccip/estimate-gas/foundry/test/SendReceive.t.sol#L53).

#### Estimate `ccipReceive` gas usage locally in your Hardhat project

To estimate the gas usage of the `ccipReceive` function within your own Hardhat project, follow these steps:

1. Create a Solidity file in the `contracts` directory of your project and import the [`MockCCIPRouter`](https://github.com/smartcontractkit/ccip/blob/ccip-develop/contracts/src/v0.8/ccip/test/mocks/MockRouter.sol) contract:

   ```solidity

>    import "@chainlink/contracts/src/v0.8/ccip/test/mocks/MockRouter.sol";

Reviewer: ```suggestion
   import {MockCCIPRouter} from "@chainlink/contracts-ccip/src/v0.8/ccip/test/mocks/MockRouter.sol";
```
